src/khepri_projection.erl

```diff
@@ -0,0 +1,296 @@
+to_ets_options(compressed, false, Acc) ->
+    Acc;
+to_ets_options(Key, Value, _Acc) ->
+    throw({unexpected_option, Key, Value}).
+
+-spec name(Projection) -> Name when
+      Projection :: projection(),
+      Name :: atom().
+%% @doc Returns the name of the given `Projection'.
+%% @end
+%% This is a utility that's necessary because the `khepri_projection'
+%% record is opaque.
+
+name(#?MODULE{name = Name}) ->
+    Name.
+
+-spec init(Projection) -> Ret when
+      Projection :: projection(),
+      Ret :: ok | {error, exists}.
+%% @hidden
+%% Initializes a projection. The current implementation creates an ETS
+%% table using the projection's `name/1' and {@link options()}.
+
+init(#?MODULE{name = Name, ets_options = EtsOptions}) ->
+    case ets:info(Name) of
+        undefined ->
+            _ = ets:new(Name, EtsOptions),
+            ok;
+        _Info ->
+            {error, exists}
+    end.
+
+-spec trigger(Projection, Path, Payload, Payload) -> Ret when
+      Projection :: projection(),
+      Path :: khepri_path:native_path(),
+      Payload :: term() | ?NO_PAYLOAD,
+      Ret :: ok.
+%% @hidden
+%% Applies the projection function against the entry from the tree to return
+%% a projected record. This projected record is then applied to the ETS table.
+
+trigger(
+  #?MODULE{name = Name, project_fun = ProjectFun},
+  Path, OldPayload0, NewPayload0) ->
+    Table = ets:whereis(Name),
+    OldPayload = ?PAYLOAD_OR(OldPayload0, undefined),
+    NewPayload = ?PAYLOAD_OR(NewPayload0, undefined),
+    try
+        case ProjectFun of
+            {simple, StandaloneFun} ->
+                trigger_simple_projection(
+                  Table, StandaloneFun, Path, OldPayload, NewPayload);
+            {extended, StandaloneFun} ->
+                Args = [Table, Path, OldPayload, NewPayload],
+                khepri_fun:exec(StandaloneFun, Args),
+                ok
+        end
+    catch
+        Class:Reason:Stacktrace ->
+            Msg = io_lib:format("Failed to trigger projection~n"
+                                "  Projection: ~s~n"
+                                "  Path:~n"
+                                "    ~p~n"
+                                "  Old payload:~n"
+                                "    ~p~n"
+                                "  New payload:~n"
+                                "    ~p~n"
+                                "  Crash:~n"
+                                "    ~ts",
+                                [Name, Path, OldPayload, NewPayload,
+                                khepri_utils:format_exception(
+                                  Class, Reason, Stacktrace, #{column => 4})]),
+            ?LOG_ERROR(Msg, []),
+            ok
+    end.
+
+-spec trigger_simple_projection(
+        Table, StandaloneFun, Path, Payload, Payload) ->
+    Ret when
+      Table :: ets:tid(),
+      StandaloneFun :: khepri_fun:standalone_fun(),
+      Path :: khepri_path:native_path(),
+      Payload :: term() | ?NO_PAYLOAD,
```

Comment:
As far as I understand, `?NO_PAYLOAD` is never passed as it is already converted to `undefined`.